x/liquidstaking/keeper/derivative.go

```diff
@@ -0,0 +1,163 @@
+			return types.ErrInvalidLiquidCoinDenom
+		}
+		// Add the liquid staking coin to the delegations
+		delegationHolder.Delegations = delegationHolder.Delegations.Add(liquidCoin)
+	}
+
+	err := k.bankKeeper.MintCoins(ctx, minttypes.ModuleName, sdk.Coins{liquidCoin})
+	if err != nil {
+		return err
+	}
+
+	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, minttypes.ModuleName, delegatorAddr, sdk.Coins{liquidCoin})
+	if err != nil {
+		return err
+	}
+
+	shares, err := k.stakingKeeper.ValidateUnbondAmount(
+		ctx, delegatorAddr, valAddr, coin.Amount,
+	)
+
+	returnAmount, err := k.stakingKeeper.Unbond(ctx, delegatorAddr, valAddr, shares)
+	if err != nil {
+		return err
+	}
+
+	if validator.IsBonded() {
+		k.bondedTokensToNotBonded(ctx, returnAmount)
+	}
+
+	// Note: UndelegateCoinsFromModuleToAccount is internally calling TrackUndelegation for vesting account
+	// TODO: sort out types.BondedPoolName vs. types.NotBondedPoolName
+	err = k.bankKeeper.UndelegateCoinsFromModuleToAccount(ctx, types.NotBondedPoolName, delegatorAddr, sdk.Coins{liquidCoin})
+	if err != nil {
+		return err
+	}
+
+	// send coins to module account
+	err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, delegatorAddr, delegationHolder.ModuleAccount.String(), sdk.Coins{liquidCoin})
+	if err != nil {
+		return err
+	}
+
+	validator, found = k.stakingKeeper.GetValidator(ctx, valAddr)
+	if !found {
+		return types.ErrNoValidatorFound
+	}
+
+	// delegate from module account
+	_, err = k.stakingKeeper.Delegate(ctx, delegationHolder.ModuleAccount, coin.Amount, stakingtypes.Unbonded, validator, true)
+	if err != nil {
+		return err
+	}
+
+	k.SetDelegationHolder(ctx, delegationHolder)
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeMintDerivative,
+			sdk.NewAttribute(sdk.AttributeKeyAmount, coin.String()),
+			sdk.NewAttribute(types.AttributeKeyValidator, delegationHolder.Validator.String()),
+		),
+	)
+
+	return nil
+}
+
+// BurnDerivative burns an existing derivative
+func (k Keeper) BurnDerivative(ctx sdk.Context, validator sdk.ValAddress, coin sdk.Coin) error {
+
+	// ctx.EventManager().EmitEvent(
+	// 	sdk.NewEvent(
+	// 		types.EventTypeMintDerivative,
+	// 		sdk.NewAttribute(sdk.AttributeKeyAmount, coin.String()),
+	// 		sdk.NewAttribute(types.AttributeKeyValidator, derivative.Validator.String()),
+	// 	),
+	// )
+
+	return nil
+}
+
+func (k Keeper) GetLiquidStakingTokenDenom(ctx sdk.Context, valAddr sdk.ValAddress) string {
+	return fmt.Sprintf("%s-%s", k.stakingKeeper.BondDenom(ctx), valAddr.String())
+}
+
+// bondedTokensToNotBonded transfers coins from the bonded to the not bonded pool within staking
+func (k Keeper) bondedTokensToNotBonded(ctx sdk.Context, tokens sdk.Int) {
+	coins := sdk.NewCoins(sdk.NewCoin(k.stakingKeeper.BondDenom(ctx), tokens))
+	// TODO: sort out types.BondedPoolName vs. types.NotBondedPoolName
+	if err := k.bankKeeper.SendCoinsFromModuleToModule(ctx, types.BondedPoolName, types.NotBondedPoolName, coins); err != nil {
```

Comment:
`BondedPoolName` and `NotBondedPoolName` can be imported from staking types.

Reply:
https://github.com/Kava-Labs/kava/pull/1271/commits/200e9785c49f4522180545dba6744b42d3691f72